### PR TITLE
SQL: Add PIVOT support

### PR DESCRIPTION
**Pull request by costin**
Add initial PIVOT support for transforming a regular table into a
statistics table around an arbitrary pivoting column:

SELECT * FROM
 (SELECT languages, country, salary, FROM mp)
 PIVOT (AVG(salary) FOR countries IN ('NL', 'DE', 'ES', 'RO', 'US'))

In the current implementation PIVOT allows only one aggregation however
this restriction is likely to be lifted in the future.

### CODE REVIEW

**Comment by costin** [comment]
@elasticmachine run elasticsearch-ci/2

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -1188,10 +1232,10 @@ protected Expression rule(Expression e) {
 
         @Override
         protected Expression rule(Expression e) {
-            if (e instanceof Alias) {
-                Alias a = (Alias) e;
-                return a.child().foldable() ? Literal.of(a.name(), a.child()) : a;
-            }
+            //            if (e instanceof Alias) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Should be removed?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    public List<NamedExpression> aggregates() {
+        return aggregates;
+    }
+    
+    public AttributeSet groupingSet() {
+        if (groupingSet == null) {
+            AttributeSet columnSet = Expressions.references(singletonList(column));
+            // grouping can happen only on "primitive" fields, thus exclude multi-fields or nested docs
+            // the verifier enforces this rule so it does not catches folks by surprise
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java]: catches -> catch

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -160,9 +161,37 @@ else if (!selectTarget.isEmpty()) {
...
                 .reduce((left, right) -> new Join(source(ctx), left, right, Join.JoinType.IMPLICIT, null))
                 .get();
+
+        // PIVOT
+        if (ctx.pivotClause() != null) {
+            PivotClauseContext pivotClause = ctx.pivotClause();
+            UnresolvedAttribute column = new UnresolvedAttribute(source(pivotClause.column), visitQualifiedName(pivotClause.column));
+            List<NamedExpression> values = namedValues(pivotClause.aggs);
+            if (values.size() > 1) {
+                throw new ParsingException(source(pivotClause.aggs), "PIVOT currently supports only one aggregation, found [{}]",
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java]: I think `ParsingException` at this point is "late", but on the other hand `SqlIllegalArgumentException` doesn't seem appropriate. Should we introduce a new one, something with 'Invalid` or `Restricted` in the name?

Not necessary to do as part of this PR though.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        if (valueOutput == null) {
+            List<Attribute> out = new ArrayList<>(aggregates.size() * values.size());
+            if (aggregates.size() == 1) {
+                NamedExpression agg = aggregates.get(0);
+                for (NamedExpression value : values) {
+                    ExpressionId id = new ExpressionId(agg.id().hashCode() + value.id().hashCode());
+                    out.add(value.toAttribute().withDataType(agg.dataType()).withId(id));
+                }
+            }
+            // for multiple args, concat the function and the value
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java]: Shouldn't we do that also for one aggregate function?

**Review by matriv — COMMENTED** [review]
Great work overall!

I left some comments and here are a couple of more general ones:
 - Should we have some test similar to https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/cli/FetchSizeTestCase.java#L56
to double check that the paging works for large row numbers
 - Could you please add some tests in `OptimizerTests` to validate the `QueryFolder` & `Optimizer` folding changes?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java]: Considering there's no typed consumer of these exceptions, the fact that it's a parsing exception vs restricted or whatever we get to call it, is about irrelevant semantics.
Further more if the client sees the message wrapped (inside the UI for example), the exception type is even further minimized.
My point is, the message sent to the user is important, not the exception in which is wrapped.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java]: No - this follows the Oracle convention which is the only one that supports multiple aggs inside the same pivot. Currently this branch is disabled due to a corner-case optimization that results in missing IDs.

**Comment by costin** [comment]
@astefan @matriv I've updated the PR, please take another look at it.

Thanks to a number of queries suggested by @astefan I concluded that supporting folding expressions inside values or functions inside columns leads to invalid queries due to the optimizer kicking in and creating different naming expressions which, due to the difference in ID, stops matching.

As such I've beefed up the Verifier and to prevent such cases - once the `NamedExpression`/`ExpressionId` refactoring will take place things, I'll revisit this.
Currently the folding of expressions leads to bugs since despite the name not changing, since most expressions are named, different ids are created leading to different identities.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 5]
```diff
@@ -844,4 +844,57 @@ public void testGeoShapeInSelect() {
...
+                error("SELECT * FROM (SELECT int, keyword, YEAR(date) FROM test) " + "PIVOT(AVG(int) FOR keyword IN ('bla'))"));
+    }
+
+    public void testPivotWithFoldableFunctionInValues() {
+        assertEquals("1:85: Non-literal [UCASE('bla')] found inside PIVOT values",
+                error("SELECT * FROM (SELECT int, keyword, bool FROM test) " + "PIVOT(AVG(int) FOR keyword IN ( UCASE('bla') ))"));
+    }
+
+    public void testPivotWithNull() {
+        assertEquals("1:85: Null not allowed as a PIVOT value",
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: @astefan Added test for forbidding `null` - it looks like it doesn't work since `In` doesn't allow it and it just removes it. Which kinda makes sense (null means the thing is missing).
This could be improved by either handling nulls as a separate filter or, preferably inside `In` directly.
@matriv thoughts?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 6]
```diff
@@ -844,4 +844,57 @@ public void testGeoShapeInSelect() {
...
+                error("SELECT * FROM (SELECT int, keyword, bool FROM test) " + "PIVOT(AVG(int) + ROUND(int) FOR keyword IN ('bla'))"));
+    }
+
+    public void testPivotValueNotFoldable() {
+        assertEquals("1:91: Non-literal [bool] found inside PIVOT values",
+                error("SELECT * FROM (SELECT int, keyword, bool FROM test) " + "PIVOT(AVG(int) FOR keyword IN ('bla', bool))"));
+    }
+
+    public void testPivotWithFunctionInput() {
+        assertEquals("1:37: No functions allowed (yet); encountered [YEAR(date)]",
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: No functions allowed due to namedExpression issues.

**Comment by costin** [comment]
Added a minor update essentially preventing `InnerAgg`s to be used. While the simple case can be addressed - `SELECT .. PIVOT (SUM_OF_SQUARES(x)...)` - the issue appears when the `InnerAgg` is hidden:
`SELECT .. PIVOT (ROUND(SUM_OF_SQUARES(x))...)`.
Again, I'm expecting the `NamedExpression` refactor to address this by using function equality instead of ids.

Until then, I've prevented the use of such aggs inside the verifier.

File: x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec [hunk 7]
```diff
@@ -152,6 +152,30 @@ SELECT * FROM (SELECT languages, gender, salary FROM test_emp) PIVOT (AVG(salary
 null           |48396.28571428572|62140.666666666664
 ;
 
+averageWithTwoValuesAndOrderDesc
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec]: AndLimit?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 8]
```diff
@@ -193,7 +193,13 @@ protected LogicalPlan rule(Pivot plan) {
                 // TODO: this should be removed when refactoring NamedExpression
                 else if (namedExpression instanceof Literal) {
                     rawValues.add(namedExpression);
-                } else {
+                }
+                // TODO: NamedExpression refactoring should remove this
+                else if (namedExpression.foldable()) {
+                    rawValues.add(Literal.of(namedExpression.name(), namedExpression));
+                }
+                // TOOD: same as above
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: typo -> TODO

File: x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec [hunk 9]
```diff
@@ -203,18 +203,4 @@ null           |31070.0
 3              |26830.0        
 4              |24646.0        
 5              |23353.0        
-;
-
-innerAggPivot
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec]: Why did you remove this? Isn't this the case that is supported?

**Review by matriv — APPROVED** [review]
LGTM. Thank you for the feedback addressing and once again great work!

File: x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec (hunk 9 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec]: https://github.com/elastic/elasticsearch/pull/46489#issuecomment-532769953

Comment (matriv) [reply x-pack/plugin/sql/qa/src/main/resources/pivot.csv-spec]: Thx, I got confused and thought that we'd support this more trivial case.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java (hunk 5 above)
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: When `null` is inside the value list of `IN` it will always yield `NULL` as a result which in the case of `IN` used as a filter in a normal WHERE clause means `false` and therefore it's eliminated. e.g.:
```
postgres=# select null in (null);
 ?column?
----------

(1 row)
```
If for pivot this case should yield `true` It should be treated differently but in my opinion not inside `IN` but somewhere else (e.g.: QueryFolder where we have the context information of IN used in a PIVOT query).
